### DISCUSSION TITLE
How to use Postgres COPY using SqlAlchemy 1.4 Async?

### DISCUSSION THREAD

**Discussion by RamonGiovane** [sqlalchemy/sqlalchemy#12062] [Usage Questions] [answered]
Hi,

I'm trying to execute a bulk insert in my application using Postgres `COPY FROM`, but I'm not sure what the best approach is since I will probably need to use the asyncpg API directly for this.

### Specs

* SQLAlchemy 1.4 + asyncpg

### Context and What I've Tried

* I'm updating a dataset in a system: a user sends updated data in a spreadsheet (with thousands of rows).
* The goal is to delete the old data and insert what the user provided.
* This task is executed in an ARQ job, which acquires a DB connection and opens a transaction using the ORM `sessionmaker`.
* The session opened here lasts for the entire job execution. That's how we have been working with ARQ and FastAPI: a single session for the entire task.
* Up next, the system fetches some data using the SQLAlchemy `session` (authenticating the user credentials, etc.).
* Next, the actual operation begins: a `TRUNCATE` is executed, followed by a `COPY FROM`.
* We open a different connection, with a separate transaction using asyncpg directly (`asyncpg.connect` and `asyncpg.transaction`) to perform the `TRUNCATE` and `COPY`.
* It's very performant, but I'm experiencing some weird behaviors, such as silent deadlocks in integration tests that mix these asyncpg API direct calls with SQLAlchemy queries executed using the ORM `session`.

I'm kind of lost.

* Should I keep opening a different connection for this, or can I use the one I already acquired with the `sessionmaker`?

I saw this old article, but I'm not sure if it's the right way to do it. Nevertheless, I couldn't reproduce it using the SQLAlchemy 1.4 Async API:

https://sorokin.engineer/posts/en/sqlalchemy_postgres_copy_from.html

#### In short: do you guys have any advice on the correct way to perform this operation using these tools?

**Comment by zzzeek** [comment, marked as answer]
first you want to get at the asyncpg connection directly using this recipe:

https://docs.sqlalchemy.org/en/20/faq/connections.html#accessing-the-underlying-connection-for-an-asyncio-driver

if you want to use the connection that you have from an ORM `Session`, replace `conn = await engine.connect()` with `conn = await session.connection()`

then use asyncpg's API directly , such as https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.copy_from_table or whichever

**Reply by RamonGiovane** [reply]
Wow, you saved me! Using the recipe you provided and executing all the steps within the same transaction already opened by SQLAlchemy worked like a charm. I haven’t experienced any more deadlocks since then.
I guess the main lesson learned here is to avoid mixing connections opened by SQLAlchemy with manually opened asyncpg connections.

Thanks for the support!
